pkg/cli/admin/mustgather/mustgather.go

```diff
@@ -492,26 +493,30 @@ func (o *MustGatherOptions) Run() error {
 	}
 	defer o.logTimestamp()
 
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
 	var wg sync.WaitGroup
 	errCh := make(chan error, len(pods))
-	podChan := make(chan *corev1.Pod)
 
-	go func() {
-		for _, pod := range pods {
-			podChan <- pod
-		}
-		close(podChan)
-	}()
+	for _, pod := range pods {
+		queue.Add(pod)
+	}
+	queue.ShutDownWithDrain()
```

Comment:
@ingvagabund @ardaguclu I'm not sure if this is the right thing to do. But without doing this, the below call wait indefinitely because the queue is 1) empty 2) not marked for shutdown:
```go
pod, quit := queue.Get()
```

Reply:
Did you try `queue.Len()` with mutexes?

Reply:
`Len()` uses locks (https://github.com/openshift/oc/blob/master/vendor/k8s.io/client-go/util/workqueue/queue.go#L187) so no need for mutex.

Reply:
```
// ShutDownWithDrain will cause q to ignore all new items added to it. As soon
// as the worker goroutines have "drained", i.e: finished processing and called
// Done on all existing items in the queue; they will be instructed to exit and
// ShutDownWithDrain will return. Hence: a strict requirement for using this is;
// your workers must ensure that Done is called on all items in the queue once
// the shut down has been initiated, if that is not the case: this will block
// indefinitely. It is, however, safe to call ShutDown after having called
// ShutDownWithDrain, as to force the queue shut down to terminate immediately
// without waiting for the drainage.
```
actually looks ok to use :) That's what we want.

Reply:
Marking resolved as it looks like we agree with the approach here.